// @aws-cdk/aws-events-targets/lib/api-destination.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiDestination = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("@aws-cdk/aws-iam");
const util_1 = require("./util");
/**
 * Use an API Destination rule target.
 */
class ApiDestination {
    constructor(apiDestination, props = {}) {
        this.apiDestination = apiDestination;
        this.props = props;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_events_targets_ApiDestinationProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.constructor);
            }
            throw error;
        }
    }
    /**
     * Returns a RuleTarget that can be used to trigger API destinations
     * from an EventBridge event.
     */
    bind(_rule, _id) {
        var _b, _c, _d, _e, _f;
        const httpParameters = ((_c = (_b = !!this.props.headerParameters) !== null && _b !== void 0 ? _b : !!this.props.pathParameterValues) !== null && _c !== void 0 ? _c : !!this.props.queryStringParameters) ? {
            headerParameters: this.props.headerParameters,
            pathParameterValues: this.props.pathParameterValues,
            queryStringParameters: this.props.queryStringParameters,
        } : undefined;
        if ((_d = this.props) === null || _d === void 0 ? void 0 : _d.deadLetterQueue) {
            util_1.addToDeadLetterQueueResourcePolicy(_rule, this.props.deadLetterQueue);
        }
        return {
            ...(this.props ? util_1.bindBaseTargetConfig(this.props) : {}),
            arn: this.apiDestination.apiDestinationArn,
            role: (_f = (_e = this.props) === null || _e === void 0 ? void 0 : _e.eventRole) !== null && _f !== void 0 ? _f : util_1.singletonEventRole(this.apiDestination, [new iam.PolicyStatement({
                    resources: [this.apiDestination.apiDestinationArn],
                    actions: ['events:InvokeApiDestination'],
                })]),
            input: this.props.event,
            targetResource: this.apiDestination,
            httpParameters,
        };
    }
}
exports.ApiDestination = ApiDestination;
_a = JSII_RTTI_SYMBOL_1;
ApiDestination[_a] = { fqn: "@aws-cdk/aws-events-targets.ApiDestination", version: "1.153.1" };
